drivers/gpio/gpio_gd32.c

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) 2021 Teslabs Engineering S.L.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gd_gd32_gpio
+
+#include <drivers/gpio.h>
+
+#include "gpio_utils.h"
+
+/** GPIO mode: input floating (CTL bits) */
+#define CTL_MODE_INP_FLOAT 0x4U
+/** GPIO mode: input with pull-up/down (CTL bits) */
+#define CTL_MODE_INP_PUPD 0x8U
+/** GPIO mode: output push-pull (CTL bits) */
+#define CTL_MODE_OUT_PP 0x8U
+/** GPIO mode: output open-drain (CTL bits) */
+#define CTL_MODE_OUT_OD 0xCU
+
+struct gpio_gd32_config {
+	struct gpio_driver_config common;
+	uint32_t reg;
+	uint32_t rcu_periph_clock;
+};
+
+struct gpio_gd32_data {
+	struct gpio_driver_data common;
+	sys_slist_t callbacks;
+};
+
+static inline int gpio_gd32_configure(const struct device *port, gpio_pin_t pin,
+				      gpio_flags_t flags)
+{
+	const struct gpio_gd32_config *config = port->config;
+
+#ifdef GD32_HAS_AF_PINMUX
```

Comment:
I think it will be better if we adopt IS_ENABLED when possible.

```
if (IS_ENABLED(GD32_HAS_AF_PINMUX)) {
...
} else {
...
}
```

Reply:
This strategy can't be used here. This implies code from both branches can be always compiled, not possible here due to non-shared registers (i.e. definitions).